active/0000-bicep-version-pinning.md

```diff
@@ -0,0 +1,135 @@
+
+----------------
+
+#### 2. Update Bicep CLI
+The Bicep CLI should be updated to check for the version constraint in the configuration.
+
+If a user runs Bicep CLI directly to build a bicep file, we will try to resolve the appropriate version constraint from the closest `bicepconfig.json`:
+    - If the resolved version matches the locally-installed Bicep version, the compilation should continue normally using the resolved version.
+    - If the resolved version constraint does not match the locally-installed Bicep version, the compilation should fail with a clear message indicating the constraint violation.
+    - If no version constraint is specified anywhere in the `bicepconfig.json` resolution chain, the compilation should fallback to using the currently installed version.
+
+
+## Drawbacks
+- **Degraded visibility into registry module version constraints:** When inspecting published module source bicep file, it wouldn't be immediately clear what version was used to compile the module just by looking at the bicep file. This is because version information would be codified in the `bicepconfig.json` which is not published with the sources. However, the user can still discern the version used with the aid of the VSCode extension; the user can navigate to the compiled JSON which contains metadata about which version was used to compile the module. 
+- **Inability to specify version override for specific individual files:** Due to how `bicepconfig.json` works, it applies to all files in the directory it is placed in. Users cannot easily override the Bicep version for individual files. This limitation can potentially be restrictive in scenarios where there may be a need to compile specific individual files with different Bicep versions for compatibility or testing purposes.
+
+## Alternatives
+### Specify Bicep version in `.bicep` files:
+This might look something like the following:
+
+```bicep
+bicep.version = '0.31.92'
+
+resource test 'Test.RP' = {...}
+```
+This approach provides more granular control over the Bicep version used for each individual file, which makes it clear which version is required for a specific file, which could be useful in scenarios where different files need to be compiled with different versions. It would also be easier to discern what version was used to compile a published module by inspecting the source.
+
+However, the downside with this approach is if usage of the same version for multiple Bicep files is desired, it would require specifying the `version` in each file which can be a maintenance overhead.
+
+## Rollout plan
+
+This proposal only involves client-side changes, and will be an opt-in feature enabled when the bicep.version configuration property is specified (essentially acting as the feature flag).
+
+## Unresolved questions
```

Comment:
It would be worth noting down what the user experience should be in VSCode if there's a mismatch. Do we need to make the user aware of this?

Reply:
If we're ok with just basic warning that the configured version doesn't match in VS code, then the implementation would be super simple. Otherwise, we'd have to find a way to decouple Bicep version from the VS code extension somehow, which will likely be very expensive.

Reply:
Added a section on this, seems basic warning is our best option here.